feat(page-header): Deprecate PageHeader (#1247)
Workday/canvas-kit: modules/docs/mdx/6.0-MIGRATION-GUIDE.mdx

import {Meta} from '@storybook/addon-docs/blocks';

<Meta title="Welcome/Migration Guides/v6.0" />

# Canvas Kit 6.0 Migration Guide

Below are the breaking changes made in Canvas Kit v6. Please
[reach out](https://github.com/Workday/canvas-kit/issues/new?labels=bug&template=bug.md) if you have
any questions about the update.

- [Codemod](#codemod)
- [Component Deprecations](#component-deprecations)
  - [PageHeader](#page-header)

## Codemod

Please use our [codemod package](https://github.com/Workday/canvas-kit/tree/master/modules/codemod)
to automatically update your code to work with a majority of the breaking changes in the migration

> from Canvas Kit v5 to v6. Simply run:

Reviewer: Shortening this a bit.
```suggestion
from Canvas Kit v5 to v6:
```

```sh
> npx @workday/canvas-kit-codemod v6 [path]
```

> Note: This codemod only works on `.js`, `.jsx`, `.ts`, and `.tsx` extensions. You may need to make
> some manual changes in other file types (`.json`, `.mdx`, `.md`, etc.).

> Note: You may need to run your linter after executing the codemod, as it's resulting formatting
> (spacing, quotes, etc.) may not match your project's styling.

**Breaking changes accounted for by this codemod will be marked with a 🤖.**

**Please verify all changes made by the codemod. As a safety precaution, we recommend committing the
changes from the codemod as a single isolated commit (separate from other changes) so you can
rollback more easily if necessary.**

[Let us know](https://github.com/Workday/canvas-kit/issues/new?labels=bug&template=bug.md) if you
encounter any issues or use cases that we've missed. The `@workday/canvas-kit-codemod` package will
help us maintain additional codemod transforms to make future migrations easier.

## Component Deprecations

### Deprecation Types

> There are two main deprecation types: soft and hard.

Reviewer: I'm assuming there aren't other types of deprecations other than soft and hard.
```suggestion
There are two types of deprecations: soft and hard.
```

#### Soft Deprecation

> During a soft deprecation, a component is still available with its full functionality, but it will
> be renamed with a prefix to indicate its soft-deprecated status. It will also include a console
> warning to announce the deprecation. This warning will only be triggered on the component's initial
> render. Soft-deprecated types or utilities that will also be renamed but will generally have no
> console warning.

Reviewer: I wasn't sure if the last sentence about soft-deprecated types and utilities just had an extra "that" or if you meant to say "that [something] will also be renamed..." -- I just assumed it was an extra "that."

```suggestion
A soft-deprecated component is still available with its full functionality, but it will have been
renamed with a prefix to indicate its soft-deprecated status. It will also include a console warning
to announce its deprecation. This warning will only be triggered on the component's initial render.

Soft-deprecated types and utilities will also be renamed but generally will not trigger a console
warning.
```

#### Hard Deprecation

> During a hard deprecation a component or package will no longer be availiable. You will need to
> follow prescribed methods to update or reach out to our team directly.

Reviewer: ```suggestion
A hard-deprecated component or package is no longer available. You will need to follow the method
prescribed in our migration guide to update your application. Please reach out to our team directly
if you need additional help.
```

### Page Header

> We are [soft deprecating](#soft-deprecation) `PageHeader`. It has been renamed to
> `DeprecatedPageHeader`. Similarly, `PageHeaderProps` has been renamed to
> `DeprecatedPageHeaderProps`. You can continue to use this component exactly the same as in v5, but
> note that we plan to [hard deprecate](#hard-deprecation) this package in Canvas Kit v7.
> 
> Below are two examples:

Reviewer: ```suggestion
We are [soft-deprecating](#soft-deprecation) `PageHeader`. It has been renamed to
`DeprecatedPageHeader`. Similarly, `PageHeaderProps` has been renamed to
`DeprecatedPageHeaderProps`. You may continue to use this component exactly as you did in v5, but
note that we plan to [hard-deprecate](#hard-deprecation) it in Canvas Kit v7.

🤖 The codemod will rename `PageHeader` and `PageHeaderProps` to their deprecated names:
```

> note that we plan to [hard-deprecate](#hard-deprecation) it in Canvas Kit v7.

Reviewer 2: We need more "why" here. We should also suggest what they should/could do when we hard deprecated it.

Author: This is a good point, and something I've talked to @jpante and @myvuuu about a bit. Most of the components being deprecated in v6 will be replaced with internal Workday components from other libraries. I know we can't link to those from public-facing docs, and I'm not sure what we're allowed to say / not say publicly.

My, Justin, and I agreed that there should be separate documentation for internal and external usage. But that doesn't exist today. I'm planning on having a follow-up task to clean up the migration guide like we did for v5 to iron out those details. But in the meantime, I'm happy to discuss this more. 👍🏻

```tsx
// v5
import PageHeader, {PageHeaderProps} from '@workday/canvas-kit-react/page-header';

export const CustomPageHeader = (props: PageHeaderProps) => {
  return <PageHeader {...props} />;
};

// v6
import DeprecatedPageHeader, {
  DeprecatedPageHeaderProps,
} from '@workday/canvas-kit-react/page-header';

export const CustomPageHeader = (props: DeprecatedPageHeaderProps) => {
  return <DeprecatedPageHeader {...props} />;
};
```

```tsx
// v5
import {PageHeader, PageHeaderProps} from '@workday/canvas-kit-react';

interface CustomPageHeaderProps extends PageHeaderProps {
  // custom page header props
}

export const CustomPageHeader = (props: CustomPageHeaderProps) => {
  return <PageHeader {...props} />;
};

// v6
import {DeprecatedPageHeader, DeprecatedPageHeaderProps} from '@workday/canvas-kit-react';

interface CustomPageHeaderProps extends DeprecatedPageHeaderProps {
  // custom page header props
}

export const CustomPageHeader = (props: CustomPageHeaderProps) => {
  return <DeprecatedPageHeader {...props} />;
};
```

> 
> 🤖 These updates will be handled automatically by the codemod.

Reviewer: ```suggestion
```
